application/src/main/java/org/thingsboard/server/service/transport/DefaultTransportApiService.java

```diff
@@ -226,6 +242,52 @@ private ListenableFuture<TransportApiResponseMsg> validateCredentials(TransportP
         }
     }
 
+    protected ListenableFuture<TransportApiResponseMsg> validateOrCreateDeviceX509Certificate(String certChain, DeviceCredentialsType credentialsType) {
+        List<String> chain = convertX509CertificateChainToList(certChain);
+        String updateDeviceCertificateValue = chain.get(0);
+        String updateDeviceCertificateHash = EncryptionUtil.getSha3Hash(updateDeviceCertificateValue);
```

Comment:
please move creation of `updateDeviceCertificateHash` and `updateDeviceCertificateValue` closer to the place, where they are going to be used. 
Somewhere near by next line:
`Device device = deviceService.findDeviceByTenantIdAndName(deviceProfile.getTenantId(), deviceName);`